examples/tv-app/tv-common/clusters/messages/MessagesManager.cpp

```diff
@@ -18,64 +18,110 @@
+                           ChipLogProgress(Zcl, "HandlePresentMessagesRequest responseID value check failed"));
+
+            VerifyOrReturn(response.label.Value().size() <= CachedMessageOption::kLabelMaxLength,
+                           ChipLogProgress(Zcl, "HandlePresentMessagesRequest label length check failed"));
+
+            CachedMessageOption option(response.messageResponseID.Value(),
+                                       std::string(response.label.Value().data(), response.label.Value().size()));
+
+            cachedMessage.AddOption(option);
+        }
+        VerifyOrReturn(iter.GetStatus() == CHIP_NO_ERROR, ChipLogProgress(Zcl, "HandlePresentMessagesRequest TLV parsing error"));
+    }
+
+    mCachedMessages.push_back(cachedMessage);
+
     // Add your code to present Message
+    ChipLogProgress(Zcl, "HandlePresentMessagesRequest complete");
 }
 
-void MessagesManager::HandleCancelMessagesRequest(const chip::app::DataModel::DecodableList<chip::ByteSpan> & messageIds)
+void MessagesManager::HandleCancelMessagesRequest(const DataModel::DecodableList<ByteSpan> & messageIds)
 {
-    // TODO: Cancel Message
+    auto iter = messageIds.begin();
+    while (iter.Next())
+    {
+        auto & id = iter.GetValue();
+
+        mCachedMessages.remove_if([id](CachedMessage & entry) {
+            if (entry.mMessageId.data_equal(id))
+            {
+                return true;
+            }
+            return false;
+        });
+        // per spec, the command succeeds even when the message id does not match an existing message
+    }
 }
 
 // Attributes
-CHIP_ERROR MessagesManager::HandleGetMessages(chip::app::AttributeValueEncoder & aEncoder)
+CHIP_ERROR MessagesManager::HandleGetMessages(AttributeValueEncoder & aEncoder)
 {
     return aEncoder.EncodeList([this](const auto & encoder) -> CHIP_ERROR {
-        for (Message & entry : mMessages)
+        for (CachedMessage & entry : mCachedMessages)
         {
-            ReturnErrorOnFailure(encoder.Encode(entry));
+            ReturnErrorOnFailure(encoder.Encode(entry.GetMessage()));
         }
         return CHIP_NO_ERROR;
     });
 }
 
-CHIP_ERROR MessagesManager::HandleGetActiveMessageIds(chip::app::AttributeValueEncoder & aEncoder)
+CHIP_ERROR MessagesManager::HandleGetActiveMessageIds(AttributeValueEncoder & aEncoder)
 {
     return aEncoder.EncodeList([this](const auto & encoder) -> CHIP_ERROR {
-        for (Message & entry : mMessages)
+        for (CachedMessage & entry : mCachedMessages)
         {
-            ReturnErrorOnFailure(encoder.Encode(entry.messageID));
+            ReturnErrorOnFailure(encoder.Encode(entry.GetMessage().messageID));
         }
         return CHIP_NO_ERROR;
     });
 }
 
 // Global Attributes
-uint32_t MessagesManager::GetFeatureMap(chip::EndpointId endpoint)
+uint32_t MessagesManager::GetFeatureMap(EndpointId endpoint)
 {
-    uint32_t featureMap = 0;
+    uint32_t featureMap = 15; // all features enabled by default
```

Comment:
Use a BitMask::Raw call and set bitmasks via constants. Numeric values are not as readable.

Reply:
addressed